### DISCUSSION TITLE
remark-stringify formatting questions

### DISCUSSION THREAD

**Discussion by rachmari** [remarkjs/remark#1166] [Q&A]
👋 Hello, I'm evaluating a couple of tools to perform linting and auto-fixing on our Markdown files, one being remark. The remark tools are intriguing to me because we already use them as part of our rendering pipeline when generating HTML from Markdown. It is also a possible avenue to upgrading our Markdown files to be CommonMark compliant by using remark-stringify to write from the AST -> Markdown.

I'm running into two issues and wondering if there are possible solutions.

1. I'm testing out stringifying some of our Markdown files and I'm finding that these type Liquid statements get escaped:

    From:
    ```
    {% data variables.contact.contact_support %}
    ```

    To: 
    ```
    {% data variables.contact.contact\_support %}
    ```

    The value `contact_support` is a variable name inside of a YAML file that we use to pull reusable data into our Markdown file at render time.

    Is there any way to configure remark-stringify to prevent escaping `_`?

2. remark-stringify prefers spaces before lists and this preference doesn't seem to be configurable. For example:

    ```
    My list:
    1. item 1
    1. item 2
    ```

    is stringified to:

    ```
    My list:

    1. item 1
    1. item 2
    ```

    While that change renders the same, [CommonMark specifies](https://spec.commonmark.org/0.30/#lists) that:
    > In CommonMark, a list can interrupt a paragraph. That is, no blank line is needed to separate a paragraph from a following list:
    ```
     Foo
     - bar
     - baz
     ```

    The reason we want to avoid remark-stringify from adding a new line is also related to our use of Liquid. For example, our Liquid may contain the first part of a numbered list:

    **{% data list-item-reusable %}**
    ```
    1. First item in the list
    ```

    **Markdown file**
    ```
    {% data list-item-reusable %}
    1. Second item in a list
    ```

    Is there a way to prevent this newline?

**Comment by wooorm** [comment]
Hi Rachael!

> upgrading our Markdown files to be CommonMark

Is the current input liquid—y old GFM-pre-CommonMark style stuff?
Next to liquid, do y’all use any other extensions?

> I'm testing out stringifying some of our Markdown files and I'm finding that these type Liquid statements get escaped:

Yeah. The thing is: we *parse* as CommonMark also, and CommonMark doesn’t support liquid, so we don’t see those variables, we just see text, and underscores, and those underscores could turn into.

What complex about liquid variables is that that’s all *preprocessed*, it’s templating. That means they could generate *arbitrary* markdown (or HTML).
Take for example `{% something %} what?`, if that would return a `*`, then that would become a list, `>` for block quote, etc. It could turn into anything! And we don’t know that.
Or a different example: `[stuff]({% url %})`, we can’t parse that, maybe it yields a URL? Or a URL and a title? Or a link closing and then another link opening so that it turns into two links?

So it’s tough to handle liquid, but there might be some ways around it, depending on your needs (here’s an old discussion: https://github.com/mapbox/retext-mapbox-standard/issues/6), let me know if you want me to delve into that.

If liquid isn’t handled, we have to escape stuff, and particularly around `_` and `*` it is very gnarly to make sure they don’t accidentally turn into emphasis, hence we do have to escape those in many cases.

> remark-stringify prefers spaces before lists and this preference doesn't seem to be configurable. For example:

It is configurable but I don’t recommend doing so.
I like the simple rule: blank lines between everything, when I author markdown.
That’s also because “interrupting” paragraphs can get complex, and everything has different rules for when that’s allowed or not.
Two cases affect lists, a) is blank lines can’t start list items when interrupting

```markdown
asd
1.
   qwe
```

asd
1.
   qwe

…and b) only `1.` is allowed to interrupt:

```markdown
asd
2. qwe
```

asd
2. qwe

So the default is to join two constructs with a blank line, but you can configure that behavior with a [`Join`](https://github.com/syntax-tree/mdast-util-to-markdown/tree/main#join).

**Comment by rachmari** [comment]
Hi @wooorm thanks for the response.

> Is the current input liquid—y old GFM-pre-CommonMark style stuff?

Yes exactly! 

> Next to liquid, do y’all use any other extensions?

Our Markdown files use frontmatter, which I've already accommodated for using `remark-frontmatter`. We have several plugins configured in our rendering pipeline to rewrite links and things like that but those are basically custom plugins on top of remark. In terms of custom Markdown extensions, Liquid is really the only one. We've created several custom Liquid extensions to accommodate additional requirements in our content.

> So it’s tough to handle liquid, but there might be some ways around it, depending on your needs (here’s an old discussion: https://github.com/mapbox/retext-mapbox-standard/issues/6), let me know if you want me to delve into that.

> If liquid isn’t handled, we have to escape stuff, and particularly around _ and * it is very gnarly to make sure they don’t accidentally turn into emphasis, hence we do have to escape those in many cases.

The example for stripping Liquid in https://github.com/mapbox/retext-mapbox-standard/issues/6 is something I considered, but I don't think that would work in our case because we want to go from Markdown -> AST -> Markdown. If we ignore the Liquid before stringifying, then the updated Markdown file would also be stripped of the Liquid. I couldn't figure out a way to ignore the Liquid but still write it out when stringifying back to a Markdown file.

I have considered using retext for prose linting, and in that case we'd modifying the AST to just mark all Liquid as inline code, which is similar to the example above. But, in those cases, we won't be auto-fixing the linting errors and don't need to rewrite the Markdown file, so that would work just fine.

> It is configurable but I don’t recommend doing so.
> I like the simple rule: blank lines between everything, when I author markdown.

I agree, I prefer the blank lines too. One solution I've entertained is making sure the list marker is not inside the Liquid conditional, but for our content writers, this would introduce a large change in the existing content and how resuables are written going forward, so I don't think that is feasible for us at the moment.  

I'll take a look at https://github.com/syntax-tree/mdast-util-to-markdown/tree/main#join. 

Thanks for the help! ✨

**Reply by wooorm** [reply]
>  If we ignore the Liquid before stringifying, then the updated Markdown file would also be stripped of the Liquid. I couldn't figure out a way to ignore the Liquid but still write it out when stringifying back to a Markdown file.

I believe that project steps around that issue, in a pretty smart way: it replaces `a {{b}} c` with `a ..... c`. You could later find all 4+ dots and replace them with the original source? Or even: replace them with an alternative such as a [directive](https://github.com/micromark/micromark-extension-directive)?

So in short, either:
1. keep existing format: read file as string -> replace braces with dots -> remark parse -> plugins -> remark stringify -> replace dots with original braces -> write file
2. change to a new format: read file as string -> replace braces with dots -> remark parse -> plugins -> replace dots with directive -> remark stringify -> write file

---

However, *assuming* the braces behave like other markdown things (i.e., like strong, or like thematic rules or so, but never as a template language), you *could* also add parser support for them.
Like https://github.com/micromark/micromark#creating-a-micromark-extension, and like https://github.com/micromark/micromark-extension-mdx-expression.

Another alternative is to do this on the AST, with a plugin right after parsing, using [`mdast-util-find-and-replace`](https://github.com/syntax-tree/mdast-util-find-and-replace) to find them and to replace them with a custom node, which you then handle (turn into a string) by passing a function in [`options.handlers`](https://github.com/syntax-tree/mdast-util-to-markdown#optionshandlers) to `remark-stringify`.
